### PR TITLE
Fixes second push after creating and pushing URL

### LINKED ISSUE #2980: Able to create multiple URL's with devfile with the same name
/kind bug

## What versions of software are you using?

**Operating System:** Linux

**Output of `odo version`:** Master

## How did you run odo exactly?

```
~/openshift/springboot  master ✗                                                                                                                                            98d ⚑ ◒  
▶ odo url create --host e8ca.engint.REDACTED.com
 ✓  URL myspring-8080.e8ca.engint.REDACTED.com created for component: myspring

To apply the URL configuration changes, please use `odo push`

~/openshift/springboot  master ✗                                                                                                                                            98d ⚑ ◒  
▶ odo url create --host e8ca.engint.REDACTED.com
 ✓  URL myspring-8080.e8ca.engint.REDACTED.com created for component: myspring

To apply the URL configuration changes, please use `odo push`

~/openshift/springboot  master ✗                                                                                                                                            98d ⚑ ◒  
▶ odo url list
Found the following URLs for component myspring
NAME              URL                                               PORT     SECURE
myspring-8080     http://myspring-8080.api.engint.REDACTED.com     8080     false
myspring-8080     http://myspring-8080.api.engint.REDACTED.com     8080     false
myspring-8080     http://myspring-8080.api.engint.REDACTED.com     8080     false

~/openshift/springboot  master ✗                                                                                                                                            98d ⚑ ◒  
▶ odo url delete myspring-8080
? Are you sure you want to delete the url myspring-8080 Yes
panic: runtime error: slice bounds out of range [3:1]

goroutine 1 [running]:
github.com/openshift/odo/pkg/envinfo.(*EnvSpecificInfo).DeleteURL(0xc000b7fe50, 0x7ffd4c13d092, 0xd, 0x0, 0x0)                                                                       
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/pkg/envinfo/envinfo.go:225 +0x4fc                                                                            
github.com/openshift/odo/pkg/odo/cli/url.(*URLDeleteOptions).Run(0xc000739b00, 0x0, 0x0)
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/pkg/odo/cli/url/delete.go:113 +0x73                                                                          
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x20aa900, 0xc000739b00, 0xc0006ce480, 0xc0007e3880, 0x1, 0x1)                                                         
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:31 +0x13c                                                              
github.com/openshift/odo/pkg/odo/cli/url.NewCmdURLDelete.func1(0xc0006ce480, 0xc0007e3880, 0x1, 0x1)                                                                                 
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/pkg/odo/cli/url/delete.go:149 +0x5e                                                                          
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute(0xc0006ce480, 0xc0007e3860, 0x1, 0x1, 0xc0006ce480, 0xc0007e3860)                                          
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702 +0x285                                                          
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc00076ed80, 0x1e94f70, 0xc00004af60, 0x0)                                                               
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783 +0x2c9                                                          
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736                                                                 
main.main()
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/cmd/odo/odo.go:75 +0x460                                                                                     


```

## Actual behavior

Error out when deleting the URL as well as being able to push a URL even though there is already one with the same name.

## Expected behavior

Being able to delete the URL as well as an error happening when trying to create a URL that has the same name.

### LINKED ISSUE #3048: Push fails second time onwards during URL apply config
/kind bug

## What versions of software are you using?

**Operating System:** Mac

**Output of `odo version`:** master

##### odo create
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo create nodejs mynode
Experimental mode is enabled, use at your own risk

Validation
 ✓  Checking devfile compatibility [113700ns]
 ✓  Validating devfile component [197572ns]

Please use `odo push` command to create the component with source deployed
```

##### odo url create
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo url create mynode --port 3000 --host <my cluster ip>.nip.io
 ✓  URL mynode created for component: mynode

To apply the URL configuration changes, please use `odo push`
```

##### odo push
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo push

Validation
 ✓  Validating the devfile [388895ns]

Creating Kubernetes resources for component mynode
 ✓  Waiting for component to start [5s]

Applying URL changes
 ✓  URL mynode: http://mynode-che.9.37.222.212.nip.io created

Syncing to component mynode
 ✓  Checking files for pushing [5ms]
 ✓  Syncing files to the component [1s]

Executing devfile commands for component mynode
 ✓  Executing devbuild command "npm install" [17s]
 ✓  Executing devrun command "nodemon app.js" [2s]

Pushing devfile component mynode
 ✓  Changes successfully pushed to component
```

##### odo second push
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo push

Validation
 ✓  Validating the devfile [97804ns]

Creating Kubernetes resources for component mynode
 ✓  Waiting for component to start [45ms]

Applying URL changes
 ✗  Failed To Update Config To Component Deployed.
```

##### odo second push
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo push

Validation
 ✓  Validating the devfile [97804ns]

Creating Kubernetes resources for component mynode
 ✓  Waiting for component to start [45ms]

Applying URL changes
 ✗  Failed To Update Config To Component Deployed.
```

## Expected behavior
should push successfully

##### cat .odo/env/env.yaml 
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ cat .odo/env/env.yaml 
ComponentSettings:
  Name: mynode
  Namespace: che
  Url:
  - Name: mynode
    Port: 3000
    host: <my cluster ip>.nip.io
    Kind: route
````
not sure if kind `route` is causing it to fail

#### preference
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo preference view
PARAMETER             CURRENT_VALUE
UpdateNotification    
NamePrefix            
Timeout               
PushTimeout           
Experimental          true
PushTarget            kube
```

/area devfile

### PR DESCRIPTION
**Pull request by mik-dass**
**What type of PR is this?**
/kind bug

**What does does this PR do / why we need it**:

Fixes second push after URL creation 

**Which issue(s) this PR fixes**:

Fixes #3048 
Fixes #2980 

**How to test changes / Special notes to the reviewer**:
- create a URL and push two times
- both pushes should pass

### CODE REVIEW

**Comment by mik-dass** [comment]
```
f \"Terraform Variables\": failed to fetch dependency of \"Bootstrap Ignition Config\": failed to fetch dependency of \"Master Machines\": failed to generate asset \"Platform Credentials Check\": validate AWS credentials: mint credentials check: error simulating policy: Throttling: Rate exceeded\n\tstatus code: 400, request id: 9160ee81-d4c2-4748-b612-99e8acf232ec"
```
/retest

**Comment by mik-dass** [comment]
```
[odo]  ✗  waited 4m0s but couldn't find running pod matching selector: 'deploymentconfig=nodejs-app'
```
/retest

**Comment by johnmcollier** [comment]
@mik-dass It seems your PR also fixes https://github.com/openshift/odo/issues/2980

**Comment by mik-dass** [comment]
@johnmcollier Thanks for the review :+1:  I have updated the PR and also modified the integration test for more validation.

File: pkg/odo/cli/url/create.go [hunk 1]
```diff
@@ -218,10 +218,8 @@ func (o *URLCreateOptions) Validate() (err error) {
 				return fmt.Errorf("the url %s already exists for a different push target, please rerun the command using a different url name", o.urlName)
 			}
 			if !pushtarget.IsPushTargetDocker() {
-				curIngressDomain := fmt.Sprintf("%v.%v", o.urlName, o.host)
-				ingressDomainEnv := fmt.Sprintf("%v.%v", localURL.Name, localURL.Host)
-				if curIngressDomain == ingressDomainEnv {
-					return fmt.Errorf("the url %s already exists", curIngressDomain)
+				if o.urlName == localURL.Name {
```
Comment (johnmcollier) [inline pkg/odo/cli/url/create.go:221]: This line fixes https://github.com/openshift/odo/issues/2980👍

File: pkg/url/url.go [hunk 2]
```diff
@@ -601,7 +602,21 @@ func Push(client *occlient.Client, kClient *kclient.Client, parameters PushParam
...
 	for urlName, urlSpec := range urlCLUSTER {
 		val, ok := urlLOCAL[urlName]
-		if !ok {
+
+		configMismatch := false
+		if ok {
+			// since the host stored in an ingress
+			// is the combination of name and host of the url
+			if val.Spec.urlKind == envinfo.INGRESS {
+				val.Spec.Host = fmt.Sprintf("%v.%v", urlName, val.Spec.Host)
```
Comment (johnmcollier) [inline pkg/url/url.go:611]: Can we get the host from `urlSpec.Spec.Host` rather than constructing it?

Comment (mik-dass) [reply pkg/url/url.go:611]: One way would be a regrex operation but that would be slower and more lines of code

**Pull request merged**